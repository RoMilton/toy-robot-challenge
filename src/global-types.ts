export type Coords = [number, number];

export enum Direction {
  North = "NORTH",
  East = "EAST",
  South = "SOUTH",
  West = "WEST",
}

export enum Error {
  CommandNotFound = "COMMAND_NOT_FOUND",
  InvalidPlaceArguments = "INVALID_PLACE_ARGUMENTS",
  RobotNotPlaced = "ROBOT_NOT_PLACED",
}
