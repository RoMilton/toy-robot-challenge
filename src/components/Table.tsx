import React from "react";
import { GRID_SIZE } from "../constants";
import { Coords, Direction } from "../global-types";
import arrow from "../arrow.png";

const cellCount = GRID_SIZE ** 2;
const cellFlexBasis = `${100 / GRID_SIZE}%`;

interface Props {
  coords?: Coords;
  direction?: Direction;
}

const getHighlightCellIndex = (coords?: Coords) =>
  coords === undefined ? undefined : coords[1] * GRID_SIZE + coords[0];

const getRotation = (direction: Direction) => {
  switch (direction) {
    case Direction.North:
      return 0;
    case Direction.East:
      return 90;
    case Direction.South:
      return 180;
    case Direction.West:
      return 270;
    default:
      return 0;
  }
};

const Table: React.FC<Props> = ({ coords, direction }) => {
  const highlightedIndex = getHighlightCellIndex(coords);
  return (
    <div className="table">
      {Array.from({ length: cellCount }).map((val, index) => (
        <div
          key={index}
          className="table-cell"
          style={{ flexBasis: cellFlexBasis }}
        >
          {coords && direction && highlightedIndex === index && (
            <img
              src={arrow}
              className="table-robot"
              alt="robot"
              style={{
                transform: `rotate(${getRotation(direction)}deg)`,
              }}
            />
          )}
        </div>
      ))}
    </div>
  );
};

export default Table;
