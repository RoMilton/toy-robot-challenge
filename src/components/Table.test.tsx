import React from "react";
import { render } from "@testing-library/react";
import Table from "./Table";
import { Direction } from "../global-types";

describe("Table", () => {
  test("renders with no robot placement", () => {
    const { asFragment } = render(<Table />);
    expect(asFragment()).toMatchSnapshot();
  });

  Object.values(Direction).forEach((dir) => {
    test(`renders Table with robot facing ${dir}`, () => {
      const { asFragment } = render(
        <Table coords={[1, 1]} direction={dir as Direction} />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
