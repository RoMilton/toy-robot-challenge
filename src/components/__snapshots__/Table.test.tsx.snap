// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders Table with robot facing EAST 1`] = `
<DocumentFragment>
  <div
    class="table"
  >
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    >
      <img
        alt="robot"
        class="table-robot"
        src="arrow.png"
        style="transform: rotate(90deg);"
      />
    </div>
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
  </div>
</DocumentFragment>
`;

exports[`Table renders Table with robot facing NORTH 1`] = `
<DocumentFragment>
  <div
    class="table"
  >
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    >
      <img
        alt="robot"
        class="table-robot"
        src="arrow.png"
        style="transform: rotate(0deg);"
      />
    </div>
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
  </div>
</DocumentFragment>
`;

exports[`Table renders Table with robot facing SOUTH 1`] = `
<DocumentFragment>
  <div
    class="table"
  >
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    >
      <img
        alt="robot"
        class="table-robot"
        src="arrow.png"
        style="transform: rotate(180deg);"
      />
    </div>
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
  </div>
</DocumentFragment>
`;

exports[`Table renders Table with robot facing WEST 1`] = `
<DocumentFragment>
  <div
    class="table"
  >
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    >
      <img
        alt="robot"
        class="table-robot"
        src="arrow.png"
        style="transform: rotate(270deg);"
      />
    </div>
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
  </div>
</DocumentFragment>
`;

exports[`Table renders with no robot placement 1`] = `
<DocumentFragment>
  <div
    class="table"
  >
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
    <div
      class="table-cell"
      style="flex-basis: 20%;"
    />
  </div>
</DocumentFragment>
`;
